### PR TITLE
Switch to SCSS: base files

### PR DESCRIPTION
**Pull request by skjnldsv**
This is the base scss files for cleanup purposes.
This is a basic css to scss conversion with a manual check using [scss-diff](https://www.npmjs.com/package/css-diff)

Commits with cleanups will follow. This is a simple pr.

**Please only use this pr to discuss issues or comments about its commits.
For arguing and/or ask question, go to the issue.**

References: #1786 & #1805 
@jancborchardt @rullzer @nickvergessen @LukasReschke @MorrisJobke @Henni @BernhardPosselt 
@nextcloud/designers

### CODE REVIEW

**Comment by Espina2** [comment]
You are changing to scss?

**Comment by skjnldsv** [comment]
> You are changing to scss?
Please see #1786

File: core/css/multiselect.scss [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+		border-top-left-radius: 8px;
+		border-top-right-radius: 8px;
+	}
+	> li {
+		overflow: hidden;
+		white-space: nowrap;
+		> input[type="checkbox"] + label:before {
+			margin-left: 7px;
+		}
+		input[type='checkbox'] {
```
Comment (skjnldsv) [inline core/css/multiselect.scss]: Will do a cleanup after. :)

File: lib/private/Template/SCSSCacher.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+		try {
+			$compiledScss = $scss->compile('@import "'.$this->fileName.'";');
+		} catch(ParserException $e) {
+			$this->logger->error($e, ['app' => 'SCSSPHP']);
+			return false;
+		}
+
+		if(file_put_contents($this->fileCache, $this->rebaseUrls($compiledScss))) {
+			$this->logger->debug($root.'/'.$file.' compiled and successfully cached', ['app' => 'SCSSPHP']);
```
Comment (Henni) [inline lib/private/Template/SCSSCacher.php]: I believe the app should be set to 'core'.
Additionally it would probably make sense to log an error if `file_put_contents` fails.

Comment (skjnldsv) [inline lib/private/Template/SCSSCacher.php]: It's going to change when I will implement appdata. But you're right, I should have done that :)

File: lib/public/ICssManager.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace OCP;
+
+/**
+ * This class provides scss functionality
+ */
+
+interface ICssManager {
+	
+	public function getCss($fileName);
```
Comment (nickvergessen) [inline lib/public/ICssManager.php]: missing `@since` on method and class

**Review by nickvergessen — DISMISSED** [review]

File: core/Controller/CssController.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
+<?php
+/**
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
+ *
+ * @author Lukas Reschke <lukas@statuscode.ch>
+ * @author Morris Jobke <hey@morrisjobke.de>
+ *
+ * @license AGPL-3.0
```
Comment (nickvergessen) [inline core/Controller/CssController.php]: This is a new file, right?
If so please replace the header

File: core/Controller/CssController.php [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+class CssController extends Controller {
+
+	/** @var ICssManager */
+	protected $cssManager;
+
+	/** @var TimeFactory */
+	protected $timeFactory;
+
+
+	public function __construct($appName, IRequest $request, ICssManager $cssManager, TimeFactory $timeFactory) {
```
Comment (nickvergessen) [inline core/Controller/CssController.php]: Docs

File: lib/private/CssManager.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	/**
+	 * CssManager constructor.
+	 *
+	 * @param IAppData $appData
+	 */
+	public function __construct(IAppData $appData) {
+		$this->appData = $appData;
+	}
+
+	public function getCss($fileName) {
```
Comment (nickvergessen) [inline lib/private/CssManager.php]: docs

File: lib/private/CssManager.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+			throw new NotFoundException();
+		}
+		try {
+			return $folder->getFile($fileName);
+		} catch(NotFoundException $e) {
+			throw new NotFoundException();
+		}
+
+	}
+}
```
Comment (nickvergessen) [inline lib/private/CssManager.php]: new line

File: lib/private/Template/CSSResourceLocator.php [hunk 8]
```diff
@@ -26,12 +26,22 @@
 namespace OC\Template;
 
 class CSSResourceLocator extends ResourceLocator {
+
+	protected $appData;
+
+	public function __construct(\OCP\ILogger $logger, $theme, $core_map, $party_map, $appData) {
```
Comment (nickvergessen) [inline lib/private/Template/CSSResourceLocator.php]: docs

File: lib/private/Template/SCSSCacher.php [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+	protected $folder;
+	protected $file;
+	protected $fileName;
+	protected $fileLoc;
+	protected $fileCache;
+	protected $rootCssLoc;
+
+	/** @var \OCP\ILogger */
+	protected $logger;
+	protected $appData;
```
Comment (nickvergessen) [inline lib/private/Template/SCSSCacher.php]: missing docs on almost all members

File: core/Controller/CssController.php (hunk 4 above)
Comment (skjnldsv) [inline core/Controller/CssController.php]: Oups! :)

File: core/Controller/CssController.php (hunk 5 above)
Comment (skjnldsv) [inline core/Controller/CssController.php]: Incoming

File: core/Controller/CssController.php [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+	 *
+	 * @param string $fileName
+	 *
+	 * @return JSONResponse
+	 */
+	public function getCss($fileName) {
+		try {
+			$cssFile = $this->cssManager->getCss($fileName);
+		} catch(NotFoundException $e) {
+			return "not found";
```
Comment (skjnldsv) [inline core/Controller/CssController.php]: Forgot that thanks

File: lib/private/Template/SCSSCacher.php [hunk 11]
```diff
@@ -108,27 +124,40 @@ private function cache() {
...
 		return preg_replace($re, $subst, $css);
 	}
 
+	/**
+	 * Return the cached css file uri
+	 * @return string
+	 */
 	public function getCachedSCSS() {
-		return $this->fileCache;
+		return "index.php/css/".$this->fileNameCSS;
```
Comment (skjnldsv) [inline lib/private/Template/SCSSCacher.php]: We should probably use a URLGenerator?

File: lib/private/Template/SCSSCacher.php [hunk 12]
```diff
@@ -108,27 +124,40 @@ private function cache() {
...
 
+	/**
+	 * Add the correct uri prefix to make uri valid again
+	 * @param string $css
+	 * @return string
+	 */
 	private function rebaseUrls($css) {
 		$re = '/url\([\'"](.*)[\'"]\)/x';
-		$subst = 'url(\'../'.$this->rootCssLoc.'/$1\')';
+		$subst = 'url(\'../../../'.$this->rootCssLoc.'/$1\')';
```
Comment (skjnldsv) [inline lib/private/Template/SCSSCacher.php]: I don't really have a better way to do that 😕

**Comment by skjnldsv** [comment]
@nextcloud/designers please review

File: lib/private/Template/SCSSCacher.php [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+		return preg_replace($re, $subst, $css);
+	}
+
+	/**
+	 * Return the cached css file uri
+	 * @return string
+	 */
+	public function getCachedSCSS() {
+		return "index.php/css/".$this->fileNameCSS;
```
Comment (nickvergessen) [inline lib/private/Template/SCSSCacher.php]: `IURLGenerator::linkToRoute('core.Css.getCss')` should work

Comment (skjnldsv) [inline lib/private/Template/SCSSCacher.php]: Thanks

**Comment by skjnldsv** [comment]
Rebased and ready to merge!

**Review by rullzer — CHANGES_REQUESTED** [review]
Can you pleae crate a PR in the 3rdparty repo for this: https://github.com/nextcloud/3rdparty

File: lib/private/Server.php [hunk 14]
```diff
@@ -406,6 +406,11 @@ public function __construct($webRoot, \OC\Config $config) {
 				$c->getConfig()
 			);
 		});
+		$this->registerService('CssManager', function (Server $c) {
+			return new CssManager(
+				$c->getAppDataDir('server')
```
Comment (rullzer) [inline lib/private/Server.php]: Please don't call this server. Call it 'css' or something. Server is way to generic.

File: lib/private/Template/SCSSCacher.php [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+	 * @return boolean
+	 */
+	public function process() {
+
+		if($this->is_cached()) {
+			return true;
+		} else {
+			return $this->cache();
+		}
+		return false;
```
Comment (rullzer) [inline lib/private/Template/SCSSCacher.php]: This is never called.

File: lib/private/Template/SCSSCacher.php [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+		try{
+			$cachedfile = $this->folder->getFile($this->fileNameCSS);
+			if( $cachedfile->getMTime() > filemtime($this->fileLoc.'/'.$this->fileNameSCSS)
+				&& $cachedfile->getSize() > 0 ) {
+				return true;
+			}
+		} catch(NotFoundException $e) {
+			return false;
+		}
+        return false;
```
Comment (rullzer) [inline lib/private/Template/SCSSCacher.php]: indentation is off

File: lib/private/Template/SCSSCacher.php [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+		$subst = 'url(\'../../../'.$this->rootCssLoc.'/$1\')';
+		return preg_replace($re, $subst, $css);
+	}
+
+	/**
+	 * Return the cached css file uri
+	 * @return string
+	 */
+	public function getCachedSCSS() {
+		return substr(\OC::$server->getURLGenerator()->linkToRoute('core.Css.getCss', array('fileName' => $this->fileNameCSS)), 1);
```
Comment (rullzer) [inline lib/private/Template/SCSSCacher.php]: Please inject the URLGenerator as well

File: lib/private/TemplateLayout.php [hunk 18]
```diff
@@ -195,7 +195,8 @@ static public function findStylesheetFiles($styles) {
 			\OC::$server->getLogger(),
 			$theme,
 			array( \OC::$SERVERROOT => \OC::$WEBROOT ),
-			array( \OC::$SERVERROOT => \OC::$WEBROOT ));
+			array( \OC::$SERVERROOT => \OC::$WEBROOT ),
+			\OC::$server->getAppDataDir('server'));
```
Comment (rullzer) [inline lib/private/TemplateLayout.php]: Same here. use something like `css`

File: lib/public/ICssManager.php [hunk 19]
```diff
@@ -0,0 +1,38 @@
+<?php
```
Comment (rullzer) [inline lib/public/ICssManager.php]: Is this file really needed? I mean once it is in OCP we have to keep it stable for a long long time. This is not useful for other apps right?

File: lib/private/CssManager.php [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OC;
```
Comment (rullzer) [inline lib/private/CssManager.php]: I think it would make sense to namespace this in `OC\CSS` or something. @nickvergessen suggetions?

File: core/Controller/CssController.php [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+	public function __construct($appName, IRequest $request, ICssManager $cssManager, TimeFactory $timeFactory) {
+		parent::__construct($appName, $request);
+
+		$this->cssManager = $cssManager;
+		$this->timeFactory = $timeFactory;
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
```
Comment (rullzer) [inline core/Controller/CssController.php]: nope! @LukasReschke

File: core/Controller/CssController.php [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+		if ($cssFile !== false) {
+			$response = new FileDisplayResponse($cssFile, Http::STATUS_OK, ['Content-Type' => 'text/css']);
+			$response->cacheFor(86400);
+			$expires = new \DateTime();
+			$expires->setTimestamp($this->timeFactory->getTime());
+			$expires->add(new \DateInterval('PT24H'));
+			$response->addHeader('Expires', $expires->format(\DateTime::RFC2822));
+			$response->addHeader('Pragma', 'cache');
+			return $response;
+		} else {
```
Comment (rullzer) [inline core/Controller/CssController.php]: Drop the else.

File: lib/private/Template/SCSSCacher.php [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+		} catch(ParserException $e) {
+			$this->logger->error($e, ['app' => 'server']);
+			return false;
+		}
+
+		try {
+			$cachedfile->putContent($this->rebaseUrls($compiledScss));
+			$this->logger->debug($this->rootCssLoc.'/'.$this->fileNameSCSS.' compiled and successfully cached', ['app' => 'server']);
+			return true;
+		} catch(NotFoundException $e) {
```
Comment (rullzer) [inline lib/private/Template/SCSSCacher.php]: putcontent doesn't throw NotFound

File: lib/private/Template/SCSSCacher.php [hunk 24]
```diff
@@ -0,0 +1,164 @@
...
+	}
+
+	/**
+	 * Cache the file with AppData
+	 * @return boolean
+	 */
+	private function cache() {
+		$scss = new Compiler();
+		$scss->setImportPaths($this->fileLoc);
+		if(\OC::$server->getSystemConfig()->getValue('debug')) {
```
Comment (rullzer) [inline lib/private/Template/SCSSCacher.php]: Inject IConfig

File: lib/private/Template/SCSSCacher.php [hunk 25]
```diff
@@ -0,0 +1,164 @@
...
+namespace OC\Template;
+
+use Leafo\ScssPhp\Compiler;
+use Leafo\ScssPhp\Exception\ParserException;
+use OCP\Files\NotFoundException;
+use OCP\IURLGenerator;
+
+class SCSSCacher {
+
+	protected $root;
```
Comment (rullzer) [inline lib/private/Template/SCSSCacher.php]: Please add PHPdoc

File: lib/private/Template/CSSResourceLocator.php [hunk 26]
```diff
@@ -26,12 +26,29 @@
 namespace OC\Template;
 
 class CSSResourceLocator extends ResourceLocator {
+
+	protected $appData;
```
Comment (rullzer) [inline lib/private/Template/CSSResourceLocator.php]: PHPdoc

File: lib/private/CssManager.php [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+	/**
+	 * Get the css file and return ISimpleFile
+	 *
+	 * @param string $fileName css filename with extension
+	 * @return ISimpleFile
+	 */
+	public function getCss($fileName) {
+		try {
+			$folder = $this->appData->getFolder('css');
+		} catch(NotFoundException $e) {
```
Comment (rullzer) [inline lib/private/CssManager.php]: No need to catch if you just throw again

File: lib/private/CssManager.php [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+			throw new NotFoundException();
+		}
+		try {
+			return $folder->getFile($fileName);
+		} catch(NotFoundException $e) {
```
Comment (rullzer) [inline lib/private/CssManager.php]: No need to catch if you just throw again

File: lib/private/CssManager.php [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+	 */
+	public function __construct(IAppData $appData) {
+		$this->appData = $appData;
+	}
+
+	/**
+	 * Get the css file and return ISimpleFile
+	 *
+	 * @param string $fileName css filename with extension
+	 * @return ISimpleFile
```
Comment (rullzer) [inline lib/private/CssManager.php]: Missing `@throws`

**Review by rullzer — DISMISSED** [review]

File: core/Controller/CssController.php (hunk 21 above)
Comment (LukasReschke) [inline core/Controller/CssController.php]: Set this to `@PublicPage` and `@NoCSRFRequired`. That's fine for this case since the data should be public. Not having PublicPage would make CSS unavailable for logged-in users.

File: lib/private/Template/CSSResourceLocator.php [hunk 30]
```diff
@@ -26,12 +26,29 @@
...
+
+	/**
+	 * @param \OCP\ILogger $logger
+	 * @param string $theme
+	 * @param array $core_map
+	 * @param array $party_map
+	 * @param IAppData $appData
+	 */
+	public function __construct(\OCP\ILogger $logger, $theme, $core_map, $party_map, $appData) {
```
Comment (LukasReschke) [inline lib/private/Template/CSSResourceLocator.php]: $coreMap and $partyMap. Use camel-case instead of underscores :)

File: core/Controller/CssController.php (hunk 21 above)
Comment (rullzer) [inline core/Controller/CssController.php]: Ah right. Forgot about public pages :P :see_no_evil:

File: lib/private/Template/CSSResourceLocator.php (hunk 30 above)
Comment (skjnldsv) [inline lib/private/Template/CSSResourceLocator.php]: I used the same as ResourceLocator 😉

File: lib/private/TemplateLayout.php (hunk 18 above)
Comment (skjnldsv) [inline lib/private/TemplateLayout.php]: So it will be css/css?

Comment (rullzer) [inline lib/private/TemplateLayout.php]: Thinking more about this. Will this stuff only be for core. Or are we planing on doing it also for apps? Because if so then it would make sense to name to folders after appid. 

@LukasReschke toughts?

Comment (skjnldsv) [inline lib/private/TemplateLayout.php]: @rullzer @LukasReschke We could do something like `appdata/css/appid`?

So for core `css/core` ?
It would be a shame not to allow apps to use the scss caching system :)

Comment (nickvergessen) [inline lib/private/TemplateLayout.php]: Well that should happen automatically right? if an app adds a scss file it gets added,if an old css file was used that one gets added...

Comment (skjnldsv) [inline lib/private/TemplateLayout.php]: Right now it's not like this. There's two types of css injection inside nextcloud:
1. Automatic search inside core/css
2. function OC\Util::add_style

The first one is only implemented right now.

File: lib/public/ICssManager.php (hunk 19 above)
Comment (skjnldsv) [inline lib/public/ICssManager.php]: Should I just remove it? I inspired myself from the avatar pr, so I figured this was needed! 😕

File: lib/private/CssManager.php (hunk 28 above)
Comment (skjnldsv) [inline lib/private/CssManager.php]: We can't remove the catch in php try iirc. 😕

**Comment by skjnldsv** [comment]
Very nice @rullzer 👍 🎉 
I need to work on the tests and rebase, then we're good :)

**Comment by skjnldsv** [comment]
Rebased and ready to review.

**Comment by skjnldsv** [comment]
We need to re-add the ParseException error when a scss fail.
And we have an issue with the installation page. We can't use scss since appdata and database isn't initialised 😕

**Comment by skjnldsv** [comment]
Script i'm using to convert css files to scss.
I was doing it manually at first. Don't forget to check for differences with  [scss-diff](https://www.npmjs.com/package/css-diff)

```
#!/bin/zsh

filename=$(basename "$1")
extension="${filename##*.}"
filename="${filename%.*}"
root=$(dirname "${1}")

scss="$root/$filename.scss"

echo "Converting $filename.$extension to $scss"

curl 'http://css2sass.herokuapp.com/' -H 'Pragma: no-cache' -H 'Origin: http://css2sass.herokuapp.com' -H 'Accept-Encoding: gzip, deflate' -H 'Accept-Language: en-us,en;q=0.5' -H 'Upgrade-Insecure-Requests: 1' -H 'Content-Type: application/x-www-form-urlencoded' -H 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8' -H 'Cache-Control: no-cache' -H 'Referer: http://css2sass.herokuapp.com/' -H 'Connection: keep-alive' --data 'page%5Bcss%5D=&page%5Bsass%5D=&commit=Convert+2+SCSS' --data-urlencode "page[css]=`cat $1`" --compressed -s |  pup '#page_sass text{}' | recode html..ascii | sed 's/  /\t/g' | sed "s/\"/'/g" > $scss
```

`find -name "*.css" -exec css2scss {} \;`

**Comment by skjnldsv** [comment]
> <skjnldsv> On a fresh nc installation
> <skjnldsv> without things set up
> <skjnldsv> The scss appdata can't work since it's using the database
> <skjnldsv> and nothing is set
> <skjnldsv> How do you suggest we serve css files in this particular case?
> <skjnldsv> What I've thought so far:
> <skjnldsv> 1. Creating a install.css file for this particular case only
> <skjnldsv> 2. serving the compiled css on-the-fly without caching until database settings are ok
> <skjnldsv> Any ideas?

I'm thinking going for the installation.css file. Much easier. But we should keep as simple as we can.
If you got any counter-arguments please say. 😄 

@jancborchardt @LukasReschke @nextcloud/designers

**Comment by jancborchardt** [comment]
@skjnldsv we need to wait with merging until we branched off stable11 anyway. :) Then we can get it in directly at the beginning of the Nextcloud 12 cycle when everything is fine

**Comment by skjnldsv** [comment]
@jancborchardt You're talking about the css files?

**Comment by jancborchardt** [comment]
@skjnldsv I mean this pull request :) Since we’re in feature freeze for Nextcloud 11 which will be released soon, this will make it into Nextcloud 12. :)

**Comment by skjnldsv** [comment]
Good! :)
There's still some work to do. Tests for example. I'd like to make some.

**Comment by rullzer** [comment]
@skjnldsv so after having another look (all that changed CSS does not make it to easy). I do have some more questions.

* Why the CSSManager? That seems to just be a dump wrapper around the AppData. No need for that if it stays the way it is now.

* So you compile each css file into the scss file right? This is done when creating a template and not when requesting a stylesheet (right?). Wouldn't it be simple to compile ondemand when it is requested from the controller?

**Comment by skjnldsv** [comment]
> Why the CSSManager? That seems to just be a dump wrapper around the AppData. No need for that if it stays the way it is now.

Because I used the overall file hierarchy nectcloud seems to use for other components (avatars...)
I figured it would be more pleasant not to change everything!

> So you compile each css file into the scss file right? This is done when creating a template and not when requesting a stylesheet (right?). Wouldn't it be simple to compile ondemand when it is requested from the controller?

This is indeed done on the first template creation. Having the compilation don on request would change the entire way nextcloud is handling css files. Like above, I dodn't wanted to change everything, and it seemed much simpler and easy to understand to use something like I did than creating an entire other system. Of course I could have created a OC\CSS part and create an independent solution, but since this is already implemented on the template generation, I kind of liked it this way ;)

File: lib/private/Template/ResourceLocator.php [hunk 31]
```diff
@@ -107,6 +107,38 @@ protected function appendIfExist($root, $file, $webRoot = null) {
...
+	 *
+	 * @param string $root path to check
+	 * @param string $file the filename
+	 * @param \OCP\Files\IAppData $appData
+	 * @param \OCP\IURLGenerator $urlGenerator
+	 * @param \OC\SystemConfig $systemConfig
+	 * @param string|null $webRoot base for path, default map $root to $webRoot
+	 * @return bool True if the resource was found and cached, false otherwise
+	 */
+	protected function cacheAndAppendScssIfExist($root, $file, $appData, $urlGenerator, $systemConfig, $webRoot = null) {
```
Comment (skjnldsv) [inline lib/private/Template/ResourceLocator.php]: Like I said on the other reply, didn't wanted to change everything 😂

**Comment by skjnldsv** [comment]
Okay, I dropped old scss converted files, and updated the new ones!
This is now or never! :)

I'm going to fix some stuff on the installation css file, but feel free to test everything asap.
Every commit on master will be a pain in the a** to fix again because of this conversion.

File: lib/private/Template/CSSResourceLocator.php [hunk 32]
```diff
@@ -25,13 +25,34 @@
 
 namespace OC\Template;
```
Comment (MorrisJobke) [inline lib/private/Template/CSSResourceLocator.php]: Please drop the executable bit on this file ;) `chmod -x lib/private/Template/CSSResourceLocator.php`

Comment (skjnldsv) [inline lib/private/Template/CSSResourceLocator.php]: I will to a quick permission fix yes

Comment (LukasReschke) [inline lib/private/Template/CSSResourceLocator.php]: > Please drop the executable bit on this file ;) chmod -x lib/private/Template/CSSResourceLocator.php
  
Time for a CI check for this? :)

File: lib/private/Template/SCSSCacher.php [hunk 33]
```diff
@@ -0,0 +1,183 @@
+<?php
```
Comment (MorrisJobke) [inline lib/private/Template/SCSSCacher.php]: Same for `lib/private/Template/ResourceLocator.php` - remove the executable bit.

**Comment by MorrisJobke** [comment]
If I check out this branch I get following in the log file:

```
{"reqId":"j9WVXoY7TSKDwU1qSMUa","remoteAddr":"::1","app":"PHP","message":"Only variables should be passed by reference at \/Users\/morrisjobke\/Projects\/nextcloud\/server\/lib\/private\/Template\/SCSSCacher.php#178","level":0,"time":"2016-12-22T14:13:02+00:00","method":"GET","url":"\/server\/index.php\/apps\/files\/?dir=\/","user":"admin","version":"12.0.0.11"}
```


And those requests return a 404:

bildschirmfoto 2016-12-22 um 15 13 18

**Comment by MorrisJobke** [comment]
With short URLs (index.php removed):

bildschirmfoto 2016-12-22 um 15 43 12
bildschirmfoto 2016-12-22 um 15 45 24

Without short URLs:

bildschirmfoto 2016-12-22 um 15 48 34

**Comment by skjnldsv** [comment]
@rullzer this seems related to the urlgenerator and the url rewriting of apache. Would you mind taking a quick look? :D

**Comment by MorrisJobke** [comment]
> @rullzer this seems related to the urlgenerator and the url rewriting of apache. Would you mind taking a quick look? :D

Ping @rullzer

**Comment by rullzer** [comment]
Somehow I missed this. This is high on my todo for this week! :)

**Comment by rullzer** [comment]
ah apache stuff... always fun. let me setup that.

**Comment by rullzer** [comment]
@MorrisJobke I can't reproduce your issue. Do you have more detailed logs for me?

File: lib/private/Template/SCSSCacher.php [hunk 34]
```diff
@@ -0,0 +1,184 @@
...
+
+	/**
+	 * Check if the file is cached or not
+	 * @param string $fileNameCSS
+	 * @param string $fileNameSCSS
+	 * @param ISimpleFolder $folder
+	 * @param string $path
+	 * @return boolean
+	 */
+	private function is_cached($fileNameCSS, $fileNameSCSS, ISimpleFolder $folder, $path) {
```
Comment (ChristophWurst) [inline lib/private/Template/SCSSCacher.php]: please rename to `isCached`

**Review by ChristophWurst — DISMISSED** [review]

**Comment by LukasReschke** [comment]
> @MorrisJobke I can't reproduce your issue. Do you have more detailed logs for me?

I can reproduce this. Let's see…

File: lib/private/Template/CSSResourceLocator.php [hunk 35]
```diff
@@ -25,13 +25,34 @@
...
+
+	/** @var SCSSCacher */
+	protected $scssCacher;
+
+	/**
+	 * @param ILogger $logger
+	 * @param string $theme
+	 * @param array $core_map
+	 * @param array $party_map
+	 * @param SCSSCacher $scssCcacher
```
Comment (LukasReschke) [inline lib/private/Template/CSSResourceLocator.php]: Typo

**Comment by LukasReschke** [comment]
https://github.com/nextcloud/server/pull/2050/commits/b2d7f75ead9043cc5a7203fbe42a9925216ce281 should fix @MorrisJobke's issue.

However, the update interface is pretty much broken. I suppose we need some hard-coded CSS there?

**Comment by MorrisJobke** [comment]
> However, the update interface is pretty much broken. I suppose we need some hard-coded CSS there?

Yes - this is because executing PHP during an upgrade is not permitted and the "this instance is currently in update mode" message is shown instead of the CSS. So we need some static CSS for this page. But this shouldn't be that hard.

**Comment by skjnldsv** [comment]
Yeah, I was merging the properties into one, but didn't finished. I will take a look.

**Comment by MorrisJobke** [comment]
Everything works except the icons. The icons.css contains a `../` too much and therefore doesn't take care of the subfolder my instance is installed on (it's installed in `localhost/server/` and the icons.css refers to `localhost/core` instead of `localhost/server/core`). @skjnldsv Could you check that one?

**Comment by skjnldsv** [comment]
@MorrisJobke can't reproduce.
The page `http://localhost:8088/server/apps/files/` load the `http://localhost:8088/server/index.php/css/core/icons.css` file which contains relative paths like this one: `.icon-download{background-image:url('../../../core/css/../img/actions/download.svg?v=1')}`

Are the css files loaded with url rewriting too on your side? That would explain the issue, but this is strange since css files are supposed to be excluded from url rewrite right?

**Comment by skjnldsv** [comment]
This is strange because on my nginx conf, the regular css files are requested without the index.php, but the scss generated css files are requested with the index.php, so it shouldn't be an issue. But I will work on it, since we know where it comes from.

**Comment by MorrisJobke** [comment]
> The page http://localhost:8088/server/apps/files/ load the http://localhost:8088/server/index.php/css/core/icons.css file which contains relative paths like this one: .icon-download{background-image:url('../../../core/css/../img/actions/download.svg?v=1')}

I don't have an `index.php` in there - maybe this is the problem then. 😕

**Comment by skjnldsv** [comment]
yep, this is the issue. I need to find a different way to calculate the different path.

**Comment by skjnldsv** [comment]
@MorrisJobke please test again

**Comment by rullzer** [comment]
* rebased 3rdparty
* rebased this on master

**Comment by MorrisJobke** [comment]
JSUnit tests fail:

```
PhantomJS 2.1.1 (Linux 0.0.0) OCA.Files.BreadCrumb tests Resizing Hides breadcrumbs to fit max allowed width FAILED
	Expected true to equal false.
	apps/files/tests/js/breadcrumbSpec.js:210:52
	Expected true to equal false.
	apps/files/tests/js/breadcrumbSpec.js:211:52
PhantomJS 2.1.1 (Linux 0.0.0) OCA.Files.BreadCrumb tests Resizing Updates the breadcrumbs when reducing max allowed width FAILED
	Expected true to equal false.
	apps/files/tests/js/breadcrumbSpec.js:239:52
	Expected true to equal false.
	apps/files/tests/js/breadcrumbSpec.js:240:52
	Expected true to equal false.
	apps/files/tests/js/breadcrumbSpec.js:241:52
PhantomJS 2.1.1 (Linux 0.0.0) OCA.Trashbin.FileList tests Breadcrumbs links the breadcrumb to the trashbin view FAILED
	Expected 3 to equal 2.
	apps/files_trashbin/tests/js/filelistSpec.js:135:34
	Expected '...' to equal 'subdir'.
	apps/files_trashbin/tests/js/filelistSpec.js:139:50
	Expected 'http://localhost/index.php/apps/files/?dir=/subdir' to equal 'http://localhost/index.php/apps/files?view=trashbin&dir=/subdir'.
	apps/files_trashbin/tests/js/filelistSpec.js:141:13
```

**Comment by MorrisJobke** [comment]
> I don't have an index.php in there - maybe this is the problem then. 😕

Yes 👍 Works now.

Now only the problem with the update page is there.

**Comment by MorrisJobke** [comment]
> JSUnit tests fail:

Let's ignore them for now - seem to be not related (maybe they are broken on master as well)

**Comment by MorrisJobke** [comment]
The list of missing files for the update page:

```
header.css
styles.css
apps.css
icons.css
multiselect.css
jquery-ui-fixes.css
tooltip.css
inputs.css
```

Let me check what I could reuse here.

**Comment by MorrisJobke** [comment]
I fixed the update page, the "this instance is in maintenance mode" page and the "this instance is too big" page with the CSS rules from the missing CSS files combined and shrinked into the update.css.

**Review by MorrisJobke — APPROVED** [review]
Tested and works here 👍

**Comment by skjnldsv** [comment]
@MorrisJobke That's why I created the installation.css file. Maybe we should merge the two of them?

**Comment by MorrisJobke** [comment]
> @MorrisJobke That's why I created the installation.css file. Maybe we should merge the two of them?

I would keep them separate, because they don't share that much rules. Most of the rules of installation are regarding input fields and for the update page it's mostly display status messages and info texts.

**Comment by rullzer** [comment]
Lets do this! :rocket:

**Review by rullzer — APPROVED** [review]

**Comment by skjnldsv** [comment]
Youhouuuuuuu

**Comment by LukasReschke** [comment]
Please fix the install page ASAP or I will revert this. Having the following in master is not an option:


screen shot 2017-01-09 at 12 29 12

**Comment by LukasReschke** [comment]
@MorrisJobke told me he is on it. I suppose the CSS links are generated wrongly:

```
					<link rel="stylesheet" href="/core/css/installation.css?v=ba222ded25d957b900c03bef914333cd">
```

My Nc is installed under a subfolder.

**Comment by MorrisJobke** [comment]
> My Nc is installed under a subfolder.

Same here.

**Comment by MorrisJobke** [comment]
> Please fix the install page ASAP or I will revert this. Having the following in master is not an option:

Fixed with #2992
